File: Controller.cpp
/*
 * ################################################################
 *
 * ProActive: The ProActive Linux Agent
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version
 * 2 of the License, or any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307
 * USA
 *
 * ################################################################
 */

/*
 * Controller.cpp
 *
 *  Created on: Mar 19, 2009
 *      Author: vasile
 */

#include "RMNodeStarter.h"
#include "CustomNodeStarter.h"
#include "P2PNodeStarter.h"
#include "NodeStarter.h"
#include "Controller.h"
#include "Constants.h"

Controller::Controller() {
	logger = log4cxx::Logger::getLogger("Controller");
}
//Controller::Controller(const Controller &controller){
//	default_node_exec = controller.default_node_exec;
//	default_shell = controller.default_shell;
////	logger =  log4cxx::Logger::getLogger("Controller");
//}
Controller::~Controller() {
}
int Controller::StartNode(string name, string java_class) {

	NodeStarter starter(name, java_class, security_policy, log4j_file,
			proactive_home, classpath, java_bin);
	starter.run();
	//	LOG4CXX_DEBUG(logger, "Node started");
	return starter.getPid();
}

int Controller::StartRMNode(string name, string java_class, string user,
		string password, string url) {

	RMNodeStarter starter(name, java_class, security_policy, log4j_file,
			proactive_home, classpath, java_bin, user, password, url);
	starter.run();
	//	LOG4CXX_DEBUG(logger, "Node started");
	return starter.getPid();
}
int Controller::StartP2PNode(string name, string java_class, string contact) {
	P2PNodeStarter starter(name, java_class, security_policy, log4j_file,
			proactive_home, classpath, java_bin, contact);
	starter.run();
	//	LOG4CXX_DEBUG(logger, "Node started");
	return starter.getPid();

}

int Controller::StartCustomNode(string name, string java_class,
		string arguments) {
	CustomNodeStarter starter(name, java_class, security_policy, log4j_file,
			proactive_home, classpath, java_bin, arguments);
	starter.run();
	//	LOG4CXX_DEBUG(logger, "Node started");
	return starter.getPid();
}

int Controller::StopNode(int pid) {
	//try to stop

	kill(pid, SIGTERM);
	//wait for the JVM to stop
	//	sleep(SLEEP_TIME_AFTER_KILL);
	bool dead = false;
	//return check for stop,
	//kill(pid, 0) returns -1 if the process does not exist
	int stop = kill(pid, 0);
	LOG4CXX_DEBUG(logger, "Kill command return after trying to stop the node " << stop);
	if (stop != 0) {
		dead = true;
	}
	//	return dead;
	return 0;
}

bool Controller::StopNode(string name) {
	LOG4CXX_ERROR(logger,"Method not implemented, use stop by PID");
	return false;
}
int Controller::SetStartConfiguration(string security_policy,
		string log4j_file, string proactive_home, string classpath,
		string java_bin) {
	this->security_policy = security_policy;
	this->log4j_file = log4j_file;
	this->proactive_home = proactive_home;
	this->classpath = classpath;
	this->java_bin = java_bin;
	return 0;
}

File: palinagent/daemon/controller.py
#! /usr/bin/env python
# -*- coding: UTF-8 -*-
#################################################################
#
# ProActive Parallel Suite(TM): The Java(TM) library for
#    Parallel, Distributed, Multi-Core Computing for
#     Enterprise Grids & Clouds
#
# Copyright (C) 1997-2010 ActiveEon 
# Contact: proactive@ow2.org or contact@activeeon.com
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; version 3 of
# the License.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this library; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307
# USA
#
# If needed, contact us to obtain a release under GPL Version 2 
# or a different license than the GPL.
#
#  Initial developer(s):               The ActiveEon Team
#  Contributor(s):
#
#################################################################
# $$ACTIVEEON_INITIAL_DEV$$
#################################################################

import logging
import threading
import time

logger = logging.getLogger("agent.ctr")


class Controller(object):
    
    def __init__(self, event_generator):
        self.evg       = event_generator.getActions()
        self.cur_event = None
        self.old_event = None
        self.must_stop = threading.Event()
        
    def start(self):
        while not self.must_stop.isSet():
            try:
                self.old_event = self.cur_event
                self.cur_event = self.evg.next()
                                
                if self.old_event is not None:
                    self.old_event.cancel()
                
                logger.debug("Controller is scheduling an event -> type:%s start:%s duration:%s" % (self.cur_event.type, self.cur_event.epoch_date, self.cur_event.duration)) 
                self.cur_event.schedule()
                
                # FIXME: sleep shouldn't be used. It should be event based ! 
                sleep_time = max(0, int((self.cur_event.epoch_date + self.cur_event.duration) - time.time()))
                logger.debug("Controller will sleep for %s seconds" % sleep_time)
                self.must_stop.wait(sleep_time)
            except StopIteration:
                logger.critical("Controller failed to get the next event for the event generator.")
                break
            except (KeyboardInterrupt, SystemExit):
                # Terminate all the processes then exit
                logger.info("Daemon exiting...")
                break
        
        # Exiting. Clean everything
        if self.old_event is not None:
            self.old_event.cancel()
        if self.cur_event is not None:
            self.cur_event.cancel()
                
    def stop(self):
        self.must_stop.set()
